### PR TITLE
Add perturbation functionality to geneformer's `.predict()` method

### PR DESCRIPTION
**Pull request by sjfleming**
Adds the optional kwargs
```
feature_deletion
feature_activation
feature_map
```
to `Geneformer.predict()`

The idea of "deletion" and "activation" are described in the Geneformer paper. Deletion is achieved by zeroing expression before tokenization. Activation is achieved by setting feature expression to a high value so that it ends up at the top of the rank-ordered list after tokenization.

"Map" is a general kind of idea to implement other kinds of _in silico_ experiments. It takes the data after tokenization, and maps tokens to other tokens. This could be used to implement the replacement of a specific token or set of tokens with a pad (0) or mask (1) token, or to switch the role of two tokens, etc.

### CODE REVIEW

**Comment by sjfleming** [comment]
Closes #98

**Comment by sjfleming** [comment]
So far there are no tests covering the new functionality

**Comment by sjfleming** [comment]
Added a test to cover the new functionality

**Comment by sjfleming** [comment]
@ordabayevy I think I might need some help with the `mypy` error here.  It's coming from my added geneformer test.  It doesn't seem to like the way I'm trying to pass **kwargs as an input?  Do you know what's wrong?  The pytest test itself passes.

**Comment by sjfleming** [comment]
Okay nevermind, I just did some extra copy and paste and now mypy seems happy.  I wonder how you're supposed to give **kwargs as an input in a way mypy will be happy with.

File: cellarium/ml/models/geneformer.py [hunk 1]
```diff
@@ -117,6 +117,42 @@ def tokenize(self, x_ng: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
...
+            assert all(
+                [g in self.var_names_g for g in feature_deletion]
+            ), "Some feature_deletion elements are not in self.var_names_g"
+            deletion_logic_g = np.logical_or.reduce([(self.var_names_g == g) for g in feature_deletion])
+            x_ng[:, deletion_logic_g] = 0
+        if feature_activation:
+            max_val = x_ng.max()
+            for i, g in enumerate(feature_activation[::-1]):
+                feature_logic_g = self.var_names_g == g
+                assert feature_logic_g.sum() == 1, f"feature_activation element {g} is not in self.var_names_g"
```
Comment (ordabayevy) [inline cellarium/ml/models/geneformer.py:139]: nit: raise `ValueError`

File: cellarium/ml/models/geneformer.py [hunk 2]
```diff
@@ -117,6 +117,42 @@ def tokenize(self, x_ng: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
...
+                x_ng[:, feature_logic_g] = top_rank_value
+
+        # tokenize and sort to give rank-ordered inputs
+        input_ids, attention_mask = self.tokenize(x_ng)
+
+        # feature map is applied after tokenization and sorting
+        if feature_map:
+            for g, target_token in feature_map.items():
+                feature_logic_g = self.var_names_g == g
+                assert feature_logic_g.sum() == 1, f"feature_map key {g} not in self.var_names_g"
```
Comment (ordabayevy) [inline cellarium/ml/models/geneformer.py:150]: nit: raise `ValueError`

File: cellarium/ml/models/geneformer.py [hunk 3]
```diff
@@ -117,6 +117,42 @@ def tokenize(self, x_ng: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
...
+                initial_token = self.feature_ids[feature_logic_g]
```
Comment (ordabayevy) [inline cellarium/ml/models/geneformer.py:151]: I wonder if there would be any benefits of having a `gene_name -> gene_id` dictionary so we don't have to do this

File: tests/test_geneformer.py [hunk 4]
```diff
@@ -67,3 +68,41 @@ def test_load_from_checkpoint_multi_device(tmp_path: Path):
...
+        case "deletion":
+            input_ids, _ = geneformer.tokenize_with_perturbations(x_ng, feature_deletion=["c"])
+            expected_input_ids = torch.tensor([[2, 3, 5, 0]])
+        case "map":
+            input_ids, _ = geneformer.tokenize_with_perturbations(x_ng, feature_map={"a": 1, "b": 1})
+            expected_input_ids = torch.tensor([[1, 1, 4, 5]])
+
+    print(f"Expected input_ids:\n{expected_input_ids}")
+    print(f"Actual input_ids:\n{input_ids}")
+    torch.testing.assert_close(input_ids, expected_input_ids)
```
Comment (ordabayevy) [inline tests/test_geneformer.py:96]: I didn't know that `torch.testing`  existed!

**Review by ordabayevy — CHANGES_REQUESTED** [review]
Looks great! Left some comments

**Comment by ordabayevy** [comment]
Notebook comment: you can directly index `CellariumPipeline`:

```py
var_names_g = pipeline.get_submodule("2").var_names_g
# instead
var_names_g = pipeline[2].var_names_g
# or better
var_names_g = pipeline.model.var_names_g
```

File: cellarium/ml/models/geneformer.py [hunk 5]
```diff
@@ -117,6 +117,42 @@ def tokenize(self, x_ng: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
...
+        feature_activation: list[str] | None = None,
+        feature_map: dict[str, int] | None = None,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+
+        # activation and deletion happen before sorting
+        if feature_deletion:
+            assert all(
+                [g in self.var_names_g for g in feature_deletion]
+            ), "Some feature_deletion elements are not in self.var_names_g"
+            deletion_logic_g = np.logical_or.reduce([(self.var_names_g == g) for g in feature_deletion])
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellarium/ml/models/geneformer.py:133]: Great call, I did not realize that!

**Review by ordabayevy — APPROVED** [review]
